### DISCUSSION TITLE
How do we use zkgroup?

### DISCUSSION THREAD

**Discussion by Atomic-Bean** [AsamK/signal-cli#380] [Q&A] [answered]
I've been trying to figure this out over the past few days and I managed to get it yesterday, but then I deleted the zkgroup file so I could rebuild a fresh one and write instructions to submit in a PR, but now signal-cli refuses to see the zkgroups stuff even when I put the same file back in the same place, I don't understand what's happening.

Here are the instructions I made, is there anything I'm missing/getting wrong?
```markdown
### Use a release build

- Go to https://github.com/signalapp/zkgroup/releases and choose the release of your choice. Download the source code and the "lib" file for your platform (`dll` for Windows, `.so` for Linux and `dylib` for MacOS)
- Place the lib file at `zkgroup/ffi/java/src/main/java/org/`
- Create a zip of the `org` directory and rename it to `zkgroup-java-[VERSION].jar` where `[VERSION]` is in the format `x.y.z`
- Place the jar in `signal-cli/src/main`

### Use the latest build

- Clone the repo and select the branch you wish to build
- Install rust if you haven't already
- Follow the "Setup" part in `OVERVIEW.md`
- Skip the gradlew step and instead make your build, the commands are
	- `make android_so	# Multiple android platform builds`
	- `make server_so	# Unknown linux GNU`
	- `make mac_dylib	# MacOS`
	- TODO: What do we do for Windows/dll and iOS?
- Your lib file will be under `zkgroup/ffi/java/src/resources/` copy it to `zkgroup/ffi/java/src/main/java/org/`
- Create a zip of the `org` directory and rename it to `zkgroup-java.jar`
- Place the jar in `signal-cli/src/main`
```

To be clear, I only ever saw it working with the latest build/master + 0.7.1 release dylib file (Because of that version mismatch, that's why I wanted to rebuild it)

The command I've been running to test if it works is `signal-cli -u +XXXXXXXXX updateProfile --name [NAME] --avatar [PATH]`

**Comment by Atomic-Bean** [comment]
I've realised there's a similar jar in `signal-cli/build/install/signal-cli/lib` for version 0.7.0 (The signal-cli script checks this file). However that jar contains an `.so` whereas I need to use `.dylib`. I tried adding the v0.7.0 `.dylib` to the jar, but I get the error `Error loading state file: org/signal/zkgroup/InvalidInputException` when running the updateProfile command. But even if it worked, I feel like that would be the incorrect way to fix this problem.

**Comment by AsamK** [comment, marked as answer]
I think the cleanest way to solve this is to remove the bundled `libzkgroup` from the `zkgroup-java-*.jar`. then java will search for the library in the system library path.

I've created a wiki page to document this, feel free to extend it ... I don't have a mac to test it:
https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal

**Reply by Atomic-Bean** [reply]
Thanks for the info, I didn't know you could do just `make libzkgroup`.

I've managed to get libzkgroup working again! Although I had an issue with one part. I don't think signal-cli sets `java.library.path` so instead it should default to the system/java path/s, however I wasn't able to figure out what my paths are. In the end I had to modify the signal-cli script itself with

```
MY_JAVA_LIBRARY_PATH="-Djava.library.path=/Users/USERNAME/Documents/libs"
exec "$JAVACMD" "$MY_JAVA_LIBRARY_PATH" "$@"
```

And that ended up working. I'd prefer to use the default `java.library.path`, but this is fine for me at least. Also I used the 0.7.0 dylib from the zkgroup releases since signal-cli's jar is v0.7.0.

Thanks again for that information and I'll update it with a few notes later today 👍

**Reply by Atomic-Bean** [reply]
Figure out how to check `java.library.path` and am now using the default instead. I've updated the wiki page with that information. Thanks again for the help and your work on signal-cli :)
